### ISSUE TITLE
Questions About `upload-recording` Behaviour

### ISSUE THREAD

**Issue by vic-ma** [rh-jmc-team/container-jfr#216]
As I was doing testing for my web PR, I noticed some maybe-problematic behaviour of `upload-recording` that I wanted to ask about (present in upstream as well as with the two pending PRs).

First, if the datasource URL is something totally invalid, like `http://abc` or `http://localhost`, then running `upload-recording` will cause CJFR to hang. I guess this is just because it waits for a response indefinitely, so maybe this is expected behaviour?

Second, if the datasource URL is pointing to a valid jfr-datasource, but it's just not the right ending, like `http://localhost:8080` (without the `/load`), then the response will be:
```
{"id":null,"commandName":"upload-recording","status":-2,"payload":"NullPointerException: "}
```

Which is sort of cryptic for what is really just an invalid URL. I ran into this one when I was mixing up adding or not adding `/load` to the env var, which I feel like may not be an uncommon problem. I haven't looked into where exactly the NPE is coming from.

One last thing is that in the web client, if I click "View in Grafana" twice fast enough (around <1 second in between), it will usually cause the web-client to disconnect (`[INFO] Disconnected remote client 192...`), and CJFR will hang afterwards (if I connect with websocat, commands don't work). [Edit: in fact, it seems like sometimes this happens even if its >1 second]

I tried replicating the same error with just the backend using websocat and I was able to do so, but I had to spam the command considerably faster than in the web client. In this case, CJFR hangs as well. I wondered if this was just a thing with commands in general, but I tried doing the same thing with the `list` command, and it seemed to work no matter how fast I spammed it.

**Comment by andrewazores** [comment]
> First, if the datasource URL is something totally invalid, like http://abc or http://localhost, then running upload-recording will cause CJFR to hang. I guess this is just because it waits for a response indefinitely, so maybe this is expected behaviour?

No, that wouldn't be expected or desirable behaviour. Any of the WebSocket Command Channel or HTTP API responses should send some form of response in all scenarios - hanging isn't the right "response" to any input. That's a bug worth filing. If the problem is that the ContainerJFR webclient instance itself is stuck waiting for a response from that invalid URL, then the request should simply have a timeout on it so that ContainerJFR eventually does reply back to its own client.

> Second, if the datasource URL is pointing to a valid jfr-datasource, but it's just not the right ending, like http://localhost:8080 (without the /load), then the response will be:
> `{"id":null,"commandName":"upload-recording","status":-2,"payload":"NullPointerException: "}`
> Which is sort of cryptic for what is really just an invalid URL. I ran into this one when I was mixing up adding or not adding /load to the env var, which I feel like may not be an uncommon problem. I haven't looked into where exactly the NPE is coming from.

Also worth filing a bug and investigating.

> One last thing is that in the web client, if I click "View in Grafana" twice fast enough (around <1 second in between), it will usually cause the web-client to disconnect ([INFO] Disconnected remote client 192...), and CJFR will hang afterwards (if I connect with websocat, commands don't work). [Edit: in fact, it seems like sometimes this happens even if its >1 second]

> I tried replicating the same error with just the backend using websocat and I was able to do so, but I had to spam the command considerably faster than in the web client. In this case, CJFR hangs as well. I wondered if this was just a thing with commands in general, but I tried doing the same thing with the list command, and it seemed to work no matter how fast I spammed it.

Not sure, I haven't seen that before, but it sounds like a nasty bug. Might be a bit tricky to trace exactly what's happening. I guess there are no logged exceptions that might reveal why the WS connection was apparently closed on the backend? Perhaps there is some problem with trying to POST the recording to the datasource and then attempting to do the same again before the first POST has completed?

**Comment by vic-ma** [comment]
>I guess there are no logged exceptions that might reveal why the WS connection was apparently closed on the backend?

Yeah, no exceptions in logs. I think more testing is needed to learn about this bug. I tried replicating it again a few times just now and in one case I couldn't get an error with a recording despite pressing the view in grafana multiple times, but when I started a new recording, the first time I tried uploading that one, it hanged. Also in these two times the logs don't say the web client disconnected like I mentioned in my post.
